Resources/public/js/comments.js

```diff
@@ -96,7 +96,7 @@
 
             event.identifier = identifier;
             event.params = {
-                permalink: encodeURIComponent(permalink || window.location.href)
+                permalink: encodeURI(permalink || window.location.href)
```

Comment:
It needs to be `encodeURIComponent`, otherwise characters like `&` and `/` won't be encoded and treated as URL.

Reply:
It's more complicated than that, as `encoreURIComponent` should be runned (as it's name imply) only against URI components and not full URIs.
I agree it can fail in some cases. The real fix would be to decompose the URL components and not use `window.location.href`.
I'll try to review and test that this week.

Reply:
The correct fix will be to revert the JS as you spotted and make something smarter in `getThreadCommentAction`.

```js
let uri = 'http://www.example.com/item-éàç-&-name/42/comments?param=valueéèçs&paramn°2=value';
encodeURIComponent(uri); // will be correct for a GET parameter but not a valid URL for URL validator
// > http%3A%2F%2Fwww.example.com%2Fitem_%C3%A9%C3%A0%C3%A7_name%2F42%2Fcomments%3Fparam%3Dvalue%C3%A9%C3%A8%C3%A7s%26paramn%C2%B02%3Dvalue
```

If we do that we will have to:
- decode on the server side,
- separate the query parts from the pathname
- re-encode them.

To get something like that :

```
http://www.example.com/item_%C3%A9%C3%A0%C3%A7_name/42/commentsparam=value%C3%A9%C3%A8%C3%A7s&paramn%C2%B02=value
```

Which is a proper URL that will pass the URLValidator test.

Reply:
In fact it does not need to be `encodeURIComponent`, because you pass the permalink to jQuery `get` method:
https://github.com/nreynis/FOSCommentBundle/blob/ef82f47be392b80eb1eeb5917207628fee8a410e/Resources/public/js/comments.js#L85

jQuery  will do the encodeURIComponent on it's own:
https://github.com/jquery/jquery/blob/master/src/serialize.js#L59

Therefore you'll have, on the client side:
```
permalink  ==>  encodeURI(permalink)  ==>  encodeURIComponent(encodeURI(permalink))
```

Then on the server side:
```
encodeURIComponent(encodeURI(permalink))  ==>  encodeURI(permalink)  ==>  permalink
                           implicitely decoded by PHP  ==> URLValidator ==> explicit urlDecode before persist
```

What you suggest will make a double encodeURIComponent which will result in something like that being submitted to the URLValidator:
```
http://www.example.com/item-éàç-name/42/comments?param=valueéèçs&paramn°2=value
http%3A%2F%2Fwww.example.com%2Fitem-%C3%A9%C3%A0%C3%A7-name%2F42%2Fcomments%3Fparam%3Dvalue%C3%A9%C3%A8%C3%A7s%26paramn%C2%B02%3Dvalue
```

Instead of something like that:
```
http://www.example.com/item-éàç-name/42/comments?param=valueéèçs&paramn°2=value
http://www.example.com/item-%C3%A9%C3%A0%C3%A7-name/42/comments?param=value%C3%A9%C3%A8%C3%A7s&paramn%C2%B02=value
```

Reply:
Copying this snippet in your browser dev tools will illustrate the point:

```
var permalink = 'http://www.example.com/item-éàç-name/42/comments?param=valueéèçs&paramn°2=value';

var p1 = encodeURI(permalink);
var p2 = encodeURIComponent(permalink);

var params = { p1: p1, p2: p2 };

jQuery.get(window.location.href, params);
// check your network inspector
console.log(params, 'check your network inspector');
```

Reply:
So bottom line, we should remove `encodeURI/encodeURIComponent` entirely from the JavaScript file?

Reply:
You can, but if you do that it becomes the responsability of `ThreadController` to escape specials characters from the `$permalink` before submitting it to the URL validator.

Both way are valid, it's up to you.

Note that if you are escaping on the server side you'll have to decompose the URL before encoding it. To my knowledge PHP lacks an equivalent of javascript's `encodeURI`. PHP's `urlencode` will encode even legal URL tokens like `:/?&=` resulting in an URIComponent and not a proper URI.

This can be done like this:
https://eval.in/947058

```php
$url_string = 'http://www.example.com/item-éàç-name/42/comments?param=valueéèçs&paramn°2=value';

$url = parse_url($url_string);
$cleanpath = implode('/', array_map(urlencode, explode('/', $url['path'])));

$query = array();
foreach(explode('&', $url['query']) as $component){
  list($key, $value) = explode('=', $component);
  $query[$key] = $value;
}
$proper_url = $url['scheme'].'://'.$url['host'].$cleanpath.'?'.http_build_query($query);

// this will pass URL Validator
echo $proper_url;
// http://www.example.com/item-%C3%A9%C3%A0%C3%A7-name/42/comments?param=value%C3%A9%C3%A8%C3%A7s&paramn%C2%B02=value

echo "\n";

// this wont
echo urlencode($url_string);
// http%3A%2F%2Fwww.example.com%2Fitem-%C3%A9%C3%A0%C3%A7-name%2F42%2Fcomments%3Fparam%3Dvalue%C3%A9%C3%A8%C3%A7s%26paramn%C2%B02%3Dvalue
```

Reply:
@nreynis What I meant is, this:

``` js
event.params = {
    permalink: encodeURIComponent(permalink || window.location.href)
}
```

Can be changed to:

``` js
event.params = {
    permalink: permalink || window.location.href
}
```

Because jQuery already runs `encodeURIComponent` on the query parameters.

Reply:
@XWB yes I understood.

My point was that if you do that you've got to do a round of escaping on the server side.

You NEED to have two layer of escaping:
* one for the transport, it's handled by jquery and automatically decoded by PHP
* a second before submitting the value to the URL Validator

Reply:
@nreynis I wasn't aware that Symfony Validator requires an encoded URL, now I understand why permalink is encoded twice in this bundle. Can you have a look at https://github.com/FriendsOfSymfony/FOSCommentBundle/pull/646, that should do the trick.

Reply:
@XWB This won't work, if you run it you'll see the validator will throw because you are submitting an URIComponent, not an URI.
You have to change the `encodeURIComponent` in `comments.js`.

Reply:
@nreynis Forgot that one, have a look: https://github.com/FriendsOfSymfony/FOSCommentBundle/pull/646/files